daemon/cmd/daemon.go

```diff
@@ -582,8 +582,73 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 	debug.RegisterStatusObject("ongoing-endpoint-creations", d.endpointCreations)
 
 	d.k8sWatcher.RunK8sServiceHandler()
+
+	if option.Config.DNSPolicyUnloadOnShutdown {
+		log.Debugf("Registering cleanup function to unload DNS policies due to --%s", option.DNSPolicyUnloadOnShutdown)
+
+		// add to pre-cleanup funcs because this needs to run on graceful shutdown, but
+		// before the relevant subystems are being shut down.
+		cleaner.preCleanupFuncs.Add(func() {
+			// Stop k8s watchers
+			log.Info("Stopping k8s service handler")
+			d.k8sWatcher.StopK8sServiceHandler()
+
+			// Iterate over the policy repository and remove L7 DNS part
+			removeL7DNSRules := func(pr policyAPI.Ports) error {
+				portProtocols := pr.GetPortProtocols()
+				if len(portProtocols) == 0 {
+					return nil
+				}
+				portRule := pr.GetPortRule()
+				if portRule == nil || portRule.Rules == nil {
+					return nil
+				}
+				dnsRules := portRule.Rules.DNS
+				log.Debugf("Found egress L7 DNS rules (portProtocol %#v): %#v", portProtocols[0], dnsRules)
+
+				// For security reasons, the L7 DNS policy must be a
+				// wildcard in order to trigger this logic.
+				// Otherwise we could invalidate the L7 security
+				// rules. This means if any of the DNS L7 rules
+				// have a matchPattern of * then it is OK to delete
+				// the L7 portion of those rules.
+				hasWildcard := false
+				for _, dns := range dnsRules {
+					if dns.MatchPattern == "*" {
+						hasWildcard = true
+						break
+					}
+				}
+				if hasWildcard {
+					portRule.Rules = nil
+				}
+				return nil
+			}
+
+			policyRepo := d.GetPolicyRepository()
+			policyRepo.Mutex.Lock()
+			policyRepo.IterateLocked(func(rule *policyAPI.Rule) {
+				for _, er := range rule.Egress {
+					_ = er.ToPorts.Iterate(removeL7DNSRules)
+				}
+			})
+			policyRepo.Mutex.Unlock()
+
+			// Bump revision to trigger policy recalculation
+			log.Infof("Triggering policy recalculation to remove DNS rules due to --%s", option.DNSPolicyUnloadOnShutdown)
```

Comment:
i am not sure how costly this regeneration is, what do think about something like below ?

```
		regenRequired := false
		cleaner.preCleanupFuncs.Add(func() {
		...
				if hasWildcard {
					portRule.Rules = nil
					regenRequired = true
				}
				return nil
                ...
                }
                ...
			if !regenRequired {
				log.Infof("Happy day, no ops due to due to --%s", option.DNSPolicyUnloadOnShutdown)
				return
			}
```

Reply:
I'm not sure I completely understand the suggestion. The regeneration needs to happen inside the pre-cleanup func, so it is performed before shutting down the agent. Also we want to make sure to iterate all possible rules. The way I read your suggestion this wouldn't be the case.

Reply:
oh sorry, properly the below diff will make it clearer

```diff
diff --git a/daemon/cmd/daemon.go b/daemon/cmd/daemon.go
index 1490322014..6d49d7809a 100644
--- a/daemon/cmd/daemon.go
+++ b/daemon/cmd/daemon.go
@@ -593,6 +593,8 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 			log.Info("Stopping k8s service handler")
 			d.k8sWatcher.StopK8sServiceHandler()
 
+			regenRequired := false
+
 			// Iterate over the policy repository and remove L7 DNS part
 			removeL7DNSRules := func(pr policyAPI.Ports) error {
 				portProtocols := pr.GetPortProtocols()
@@ -621,6 +623,7 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 				}
 				if hasWildcard {
 					portRule.Rules = nil
+					regenRequired = true
 				}
 				return nil
 			}
@@ -634,6 +637,11 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 			})
 			policyRepo.Mutex.Unlock()
 
+			if !regenRequired {
+				log.Infof("Happy day, no ops due to due to --%s", option.DNSPolicyUnloadOnShutdown)
+				return
+			}
+
 			// Bump revision to trigger policy recalculation
 			log.Infof("Triggering policy recalculation to remove DNS rules due to --%s", option.DNSPolicyUnloadOnShutdown)
 			policyRepo.BumpRevision()

```

Reply:
Thanks for the diff @sayboras! Agree, there is no need to trigger a policy regeneration in case of no DNS rules. Will change accordingly.